keps/sig-api-machinery/20200421-static-admission-webhook-configuration.md

```diff
@@ -0,0 +1,149 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Static Webhook configuration allows registering admission webhooks during
+kube-apiserver start up allowing for no delays in policy enforcement between
+policy addition and kube-apiserver startup. 
+
+## Motivation
+
+Today most policy enforcement is implemented through MutatingAdmissionWebhooks
+and/or ValidatingAdmissionWebhooks. These admission webhooks are registered
+through creating MutatingWebhookConfiguration or ValidatingWebhookConfiguration
+objects. Any policy enforcement is not in place until these webhook
+configurations are created, thereby registering the webhook. This creates a gap
+in enforcement spanning from when the kube-apiserver is started to until the
+webhook configuration objects are created and picked up by the dynamic admission
+controller. Another gap is the inability of the cluster administrator to protect
+against deletion of these webhook configuration objects as
+MutatingWebhookConfiguration and ValidatingWebhookConfiguration objects are not
+subject to webhook admission policy.
+
+This KEP aims to address these issues.  
+
+### Goals
+
+- A robust admission webhook registration process where there is no period of
+  time between the kube-apiserver coming up and the registration of an admission
+  webhook.
+
+- The registration of webhooks registered through this process should be
+  protected from alteration by API requests made by cluster users.
+
+- The new registration process should also have a highly privileged means of
+  being able to alter webhook registrations.
+ 
+## Proposal
+
+In short, this proposal is about augmenting `AdmissionConfig`'s plugin
+`configuration` [specification](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#authenticate-apiservers)
+to include a path to a configuration file containing one of the following:
+- `admissionregistration.k8s.io/v1.ValidatingWebhookConfigurationList` 
+- `admissionregistration.k8s.io/v1.MutatingWebhookConfigurationList` 
+- v1.List with items of admissionregistration.k8s.io/v1.ValidatingWebhookConfiguration 
+- v1.List  with items of admissionregistration.k8s.io/v1.MutatingWebhookConfiguration.
+This configuration is loaded  as static webhooks in api servers. These
+webhooks are called by validating and mutating admission plugins along with
+dynamically loaded webhooks for relevant admission requests. This also means
```

Comment:
I'd like to see more discussion about the implications of this bit:

1. are webhooks in the manifest file required to have names? unique names (within the file)? (I would expect yes to both)
2. what happens if a webhook in the manifest file and an API webhook object have the same name? (I would expect both to be invoked, as they are effectively in different domains)
3. would metrics and audit annotations from these webhooks be surfaced identically or differently (see admissionmetrics.Metrics calls)? it is likely that config-file-based webhooks would be monitored differently, since they are fully under the control of the kube-apiserver admin and any issues with them can only be resolved by the admin

Reply:
1. Yes to both. Added more info.
2. Both will be invokes. Added more info.
3. Good point. added more info.